### PR TITLE
feat: allow app environment values to contain spaces

### PR DESCRIPTION
**Pull request by thalesmg**
When one needs to set an application environment value that contains
spaces, such as a Mnesia directory path with spaces, the current
`--erl` switch does not pass the spaces correctly down to `erl`:

```
ͳ bin/iex --erl "-mnesia dir \"/tmp/abc/with spaces/\"" -e ':init.get_arguments()[:mnesia] |> IO.inspect(); System.halt()'
Erlang/OTP 24 [erts-12.1.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit]

['dir', '"/tmp/abc/with', 'spaces/"']
```

Which fails when loading the app:

```
iex(1)> Application.load(:mnesia)

13:26:44.461 [error] application_controller: unterminated string starting with "/tmp/abc/with": "/tmp/abc/with

{:error, {:bad_environment_value, '"/tmp/abc/with'}}
```

This adds a new switch, `--app-env`, that helps with setting such
values while preserving both the double quotes needed for Erlang to
parse the value as a string and the spaces that might be in such
string.

```
ͳ bin/iex --app-env mnesia dir "\"/tmp/abc/with spaces/\"" -e ':init.get_arguments()[:mnesia] |> IO.inspect(); System.halt()'
Erlang/OTP 24 [erts-12.1.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit]

['dir', '"/tmp/abc/with spaces/"']
```

### CODE REVIEW

**Comment by josevalim** [comment]
Is this something we could fix instead? 🤔

**Comment by thalesmg** [comment]
> Is this something we could fix instead? thinking

Good question :thinking: 

I tried to fiddle with `--erl` quoting/escaping in several ways, but couldn't find a way that preserves double quotes in particular, which are needed for setting string values such as `-mnesia dir`.

The closest I got was something like this:

```sh
erl_set () {
  eval "E${E}=\$@"
  E=$((E + 1))
}

handle_args () {
  eval 'for x in '$@'; do erl_set "$x"; done'
}

# ...

    --erl)
        S=2
        handle_args "$2"
        ;;
```

While this preserves the spaces correctly, it also eats the double quotes.  :crying_cat_face:

**Comment by josevalim** [comment]
I would if specifying --erl multiple times could work `--erl "-mnesia dir" --erl "foo bar"? Or we provide some sort of --unquoted-erl tag.

**Comment by thalesmg** [comment]
> I wonder if specifying --erl multiple times could work `--erl "-mnesia dir" --erl "foo bar"? Or we provide some sort of --unquoted-erl tag.

Unfortunately, the first alternative doesn't appear to regarding the spaces:

```sh
ͳ bin/iex --erl "-mnesia dir" --erl "\"/tmp/abc/with spaces/\"" -e ':init.get_arguments()[:mnesia] |> IO.inspect(); System.halt()'
Erlang/OTP 24 [erts-12.1.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit]

['dir', '"/tmp/abc/with', 'spaces/"']
```

The `--unquoted-erl` would perhaps entail appending the arguments at the [last moment](https://github.com/elixir-lang/elixir/blob/cb0534bfa05f9651e81f3ec7171725de2475d616/bin/elixir#L240) using `set --`?  I'll try that.

**Comment by thalesmg** [comment]
I couldn't figure out how to do the escaping for `--unquoted-erl` (which was basically my first approach :sweat_smile: ), it keeps breaking the spaces whenever I get a "working" version.
If you have an idea of how to do it, I'd appreciate.  :smile_cat:

**Comment by josevalim** [comment]
Ok, let's go with the approach in the PR then. Some notes:

1. You need to add a test. Check kernel/cli_test.ex
2. You have to make it work on Windows too

Thanks!

File: bin/elixir [hunk 1]
```diff
@@ -20,6 +20,7 @@ Usage: $(basename "$0") [options] [.exs file] [data]
 
   --app APP                    Starts the given app and its dependencies (*)
   --erl "SWITCHES"             Switches to be passed down to Erlang (*)
+  --app-env APP KEY VAL        Sets app environment variable (*)
```
Comment (josevalim) [inline bin/elixir:23]:
```suggestion
  --app-env APP KEY VAL        Sets app environment variable (*)
    --erl "SWITCHES"             Switches to be passed down to Erlang (*)
```

File: bin/elixir [hunk 2]
```diff
@@ -29,6 +30,7 @@ Usage: $(basename "$0") [options] [.exs file] [data]
 
 Options given after the .exs file or -- are passed down to the executed code.
 Options can be passed to the Erlang runtime using \$ELIXIR_ERL_OPTIONS or --erl.
+If an application environment value contains spaces, use --app-erl instead of --erl.
```
Comment (josevalim) [inline bin/elixir:33]:
```suggestion
```

**Comment by josevalim** [comment]
Actually, there is an issue in supporting --app-env as is today. It currently expects terms to be given in Erlang syntax, which would be confusing to do through Elixir's command line. And I don't think there is any trivial fix for this.

**Comment by thalesmg** [comment]
> Actually, there is an issue in supporting --app-env as is today. It currently expects terms to be given in Erlang syntax, which would be confusing to do through Elixir's command line. And I don't think there is any trivial fix for this.

Indeed, that could be confusing.  Maybe the arguments would have to be parsed by Elixir before being fed to the Erlang VM, but certainly not trivial.  :thinking:

File: bin/elixir.bat [hunk 3]
```diff
@@ -142,6 +143,7 @@ if ""==!par:--no-halt=!   (set "parsElixir=!parsElixir! --no-halt" && goto start
 if ""==!par:--remsh=!     (set "parsElixir=!parsElixir! --remsh %1" && shift && goto startloop)
 if ""==!par:--dot-iex=!   (set "parsElixir=!parsElixir! --dot-iex %1" && shift && goto startloop)
 rem ******* ERLANG PARAMETERS **********************
+if ""==!par:--app-env=!         (set "parsErlang=!parsErlang! -%1 %2 %3" && shift && shift && shift && goto startloop)
```
Comment (thalesmg) [inline bin/elixir.bat:146]: For some bizarre reason, if I simply add more spaces to align this, the script breaks with `The system cannot find the batch label specified expand_erl_libs` :hear_no_evil: :thinking:

**Comment by josevalim** [comment]
Btw, would setting ERL_AFLAGS or ERL_ZFLAGS also work? Or do parsing also get in the way?

In any case, I don't think there is something we can do in Elixir CLI itself, so I will go ahead and close this. :)

**Comment by josevalim** [comment]
Ah, smart call on the vm.args too!

**Pull request closed without merging**